recipes-app/iot2050-firmware-update/files/iot2050-firmware-update.py

```diff
@@ -0,0 +1,183 @@
+
+        TMP_CREATE_CMD=["mktemp","-p","/dev/shm"]
+        TMP_IMG=self.shell_cmd_execute(TMP_CREATE_CMD).stdout.decode()
+        TMP_MTD=self.shell_cmd_execute(TMP_CREATE_CMD).stdout.decode()
+        MTDNUM=0
+        IMGPOS=0
+        SKIP_BYTES=0
+
+        print("===================================================")
+        print("IOT2050 firmware update started - DO NOT INTERRUPT!")
+        print("===================================================")
+        while True:
+            MTD_SYS="/sys/bus/platform/devices/47040000.spi/mtd/mtd{}".format(MTDNUM)
+            GET_MTD_SIZE_CMD=["cat","{}/size".format(MTD_SYS)]
+            GET_MTD_ERASESIZE_CMD=["cat","{}/erasesize".format(MTD_SYS)]
+            GET_MTD_NAME=["cat","{}/name".format(MTD_SYS)]
+
+            if not os.path.exists(MTD_SYS):
+                raise FirmwareUpdateException("{} not Exists".format(MTD_SYS))
+
+            MTD_SIZE=self.shell_cmd_execute(GET_MTD_SIZE_CMD).stdout.decode()
+            MTD_ERASESIZE=self.shell_cmd_execute(GET_MTD_ERASESIZE_CMD).stdout.decode()
+            MTD_SIZE=int(MTD_SIZE)
+            MTD_ERASESIZE=int(MTD_ERASESIZE)
+
+            SKIP=False
+            mtd_name = self.shell_cmd_execute(GET_MTD_NAME).stdout.decode()
+            if not self.erase_env:
+                if mtd_name.strip() == "ospi.env" or mtd_name.strip()=="ospi.env.backup":
+                    MTDNUM=MTDNUM+1
+                    SKIP=True
+            if SKIP:
+                SKIP_BYTES=SKIP_BYTES+MTD_SIZE
+                print("Skip Updating %-20s" % mtd_name)
+                continue
+            else:
+                print("Updating %-20s" % mtd_name.strip(),end="")
+            MTDPOS=0
+            while MTDPOS < MTD_SIZE:
+                TMP_CLEAR_CMD=["rm","-f","{0}".format(TMP_IMG)]
+                self.shell_cmd_execute(TMP_CLEAR_CMD)
+                self.dd_execute(filename,IMGPOS+SKIP_BYTES,MTD_ERASESIZE,TMP_IMG,0)
+                if not os.path.getsize(TMP_IMG):
+                    print("\n<end of image>")
+                    CLEAR_TMP_CMD=["rm","-f","{}".format(TMP_IMG),"{}".format(TMP_MTD)]
+                    self.shell_cmd_execute(CLEAR_TMP_CMD)
+                    print("Completed. Please reboot the device")
+                    return
+                self.dd_execute("/dev/mtd{}".format(MTDNUM),MTDPOS,MTD_ERASESIZE,TMP_MTD,0)
+                if not filecmp.cmp(TMP_IMG,TMP_MTD):
+                    print("U",end="")
+                    ERASE_CMD=["flash_erase","-q","/dev/mtd{}".format(MTDNUM),"{}".format(MTDPOS),"1"]
+                    self.shell_cmd_execute(ERASE_CMD)
+                    self.dd_execute(TMP_IMG,0,MTD_ERASESIZE,"/dev/mtd{}".format(MTDNUM),MTDPOS,MTD_ERASESIZE)
+                else:
+                    print(".",end="")
+                IMGPOS=IMGPOS+MTD_ERASESIZE
+                MTDPOS=MTDPOS+MTD_ERASESIZE
+            MTDNUM=(MTDNUM+1)
+            print("\n")
```

Comment:
This should also be rewritten to use Python were possible. The current code is basically wrapping the a shell script into Python, rather than expressing the task in Python directly.

There is even a way to request the flash erasing natively (MEMERASE ioctl for mtd), which would have the benefit of overcoming the tools dependency.